docs/high-performance/MemoryOwner.md

```diff
@@ -0,0 +1,72 @@
+keywords: windows 10, uwp, windows community toolkit, uwp community toolkit, uwp toolkit, parallel, high performance, net core, net standard
+dev_langs:
+  - csharp
+---
+
+# MemoryOwner&lt;T>
+
+The [MemoryOwner&lt;T>](https://docs.microsoft.com/dotnet/api/microsoft.toolkit.highperformance.buffers.memoryowner-1) is a buffer type implementing `IMemoryOwner<T>`, an embedded length property and a series of performance oriented APIs. It is essentially a lightweight wrapper around the `ArrayPool<T>` type, with some additional helper utilities.
+
+## How it works
+
+`MemoryOwner<T>` has the following main features:
+
+- One of the main issues of arrays returned by the `ArrayPool<T>` APIs and of the `IMemoryOwner<T>` instances returned by the `MemoryPool<T>` APIs is that the size specified by the user is only being used as a _minum_ size: the actual size of the returned buffers might actually be greater. `MemoryOwner<T>` solves this by also storing the original requested size, so that `Memory<T>` and `Span<T>` instances retrieved from it will never need to be manually sliced.
+- When using `IMemoryOwner<T>`, getting a `Span<T>` for the underlying buffer requires first to get a `Memory<T>` instance, and then a `Span<T>`. This is fairly expensive, and often unnecessary, as the intermediate `Memory<T>` might actually not be needed at all. `MemoryOwner<T>` instead has an additional `Span` property which is extremely lightweight, as it directly wraps the internal `T[]` array being rented from the pool.
+- Buffers rented from the pool are not cleared by default, which means that if they were not cleared when being previous returned to the pool, they might contain garbage data. Normally, users are required to clear these rented buffers manually, which can be verbose especially when done frequently. `MemoryOwner<T>` has a more flexible approach to this, through the `Allocate(int, AllocationMode)` API. This method not only allocates a new instance of exactly the requested size, but can also be used to specify which allocation mode to use: either the same one as `ArrayPool<T>`, or one that automatically clears the rented buffer.
+- There are cases where a buffer might be rented with a greater size than what is actually needed, and then resized afterwards. This would normally require users to rent a new buffer and copy the region of interest from the old buffer. Instead, `MemoryOwner<T>` exposes a `Slice(int, int)` API that simply return a new instance wrapping the specified area of interest. This allows to skip renting a new buffer and copying the items entirely.
+
+## Syntax
+
+Here is an example of how to rent a buffer and retrieve a `Memory<T>` instance:
+
+```csharp
+// Be sure to include this using at the top of the file:
+using Microsoft.Toolkit.HighPerformance.Buffers;
+
+using (MemoryOwner<int> buffer = MemoryOwner<int>.Allocate(42))
+{
+    // Buffer has exactly 42 items
+    Memory<int> memory = buffer.Memory;
+    Span<int> span = buffer.Span;
```

Comment:
Maybe add to this example showing how to write/read from the buffer?

Reply:
Done in https://github.com/MicrosoftDocs/WindowsCommunityToolkitDocs/pull/301/commits/1a57734bd1d853d4b44364a467af8f4d2cc0ba44.